compiler/rustc_mir_build/src/build/matches/test.rs

```diff
@@ -438,12 +442,51 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
             }
         }
 
-        let ty::Ref(_, deref_ty, _) = *ty.kind() else {
-            bug!("non_scalar_compare called on non-reference type: {}", ty);
-        };
+        match *ty.kind() {
+            ty::Ref(_, deref_ty, _) => ty = deref_ty,
+            _ => {
+                ty = ty.fold_with(&mut BottomUpFolder {
+                    tcx: self.tcx,
+                    ty_op: |ty| match *ty.kind() {
+                        ty::FnPtr(sig) => {
+                            // PartialEq is not implemented for `for<'a> fn(&'a ())`, so
+                            // we convert the types a bit to make them work out.
```

Comment:
i am not yet convinced that this is necessary (and it is definitely theoretically unsound)

i can imagine that we may have to walk types which have builtin structural equality and replace higher ranked fn ptrs there, but I definitely don't want to replace fn ptrs inside of adts

Reply:
You mean in the substs of Adts? That is exactly the use case and the reason I made this change. I'll dig up the issue with previous discussion on the topic.